### DISCUSSION TITLE
What is the method to eliminate left recursion in Chumsky?

### DISCUSSION THREAD

**Discussion by chiichen** [zesterer/chumsky#575] [General]
I am very interested in what method is used to eliminate left recursion in Chumsky. Is it the `packrat` algorithm or trying to rewrite the left recursive grammar into right recursion?

**Comment by zesterer** [comment]
Chumsky isn't capable of eliminating it by itself. You broadly have two options:

1) Rewrite the grammar to be right-recursive instead

2) Use the `.memoised()` combinator (this is less preferable since it comes with overhead, but it is occasionally necessary)

In almost all cases, (1) suffices. For example, traditional BNF contains only recursion and no iteration, but in chumsky you can use `.repeated()` (and friends) to easy express many left-recursive patterns with iteration instead. For example,

```rs
// Parse `a.b.c.d`
let expr = recursive(|expr| expr.then_ignore(just('.')).then(expr).map(|(_, _)| ...));
```

Can be rephrased as

```rs
let expr = ... ;
let access_expr = expr.foldl(
    just('.').ignore_then(expr).repeated(),
    |_, _| ... ,
);
```

There are many more examples of this sort of thing both in the crate examples and the inline docs, which have many fully-fledged example parsers.

Also, the new pratt parsing API means expressing many of these patterns *much* simpler: https://docs.rs/chumsky/1.0.0-alpha.6/chumsky/pratt/index.html

**Reply by chiichen** [reply]
so recursive closure just simply express the recursive use of the expression itself without any other implicite algorithm inside?

**Reply by Zij-IT** [reply]
Correct. Recursive gives you the ability to refer to an element within its own definition.

Thus allowing:

```
expr := '(' expr (',' expr)* ')'
      | num
```

Note, that this does not allow for implicit or explicit left recursion, such as:

```
-- Explicit
expr := expr '+' expr
      | num

-- Implicit
expr := other '+' expr
      | num

other := expr '^' expr
       | num
```

Left recursion is recognizable where an item is the left most symbol in its own definition. In the implicit case, there is left recursion because by substitutnt `other`, we get `expr` as a leftmost symbol

You can resolve left recursion by either using precedence climbing, Pratt parsing, or by restructuring the grammar. Of those, Pratt parsing is directly supported by chumsky through the `Parser::pratt` method.

**Reply by chiichen** [reply]
thanks a lot!
